# Предобработка данных

Отработка алгоритма решения задачи предобработки данных. 

Датасет <code>dtp_last.csv</code> содержит информацию о ДТП в Российской Федерации в 2018 году.

<li>'reg_code' - код региона (октмо)
<li>'регион' - название региона
<li>'road_code' - код дороги
<li>'Дорога' -  название дороги
<li>'road_type' - тип дороги
<li>'address' -  адрес
<li>'crash_type_name' -тип ДТП
<li>'crash_date' - дата ДТП
<li>'crash_time' - время ДТП
<li>'crash reason' - причина ДТП
<li>'fatalities_amount' - число погибших
<li>'victims_amount' - число пострадавших
<li>'vehicles_amount' - число транспортных средств
<li>'participants_amount' - число участников



Задача: Подготовить данные для дальнейшего исследования.
1. Изучите общую информацию.
2. Осуществите предобработку заголовков.
3. Заполните пропуски.
4. Найдите проблемы в данных.
5. Преобразуйте типы данных. Добавьте столбец datetime из имеющихся столбцов, используя разделитель T. Преобразуйте к типу данных "дата/время".
6. Удалите дубликаты.
7. Добавьте столбец с днями недели на русском языке.
8. Экспортируйте результат в excel.
9. Сформулируйте общий вывод по работе.

Работу сопроводите выводами и комментариями к коду.

### Загрузка данных. Общая информация

In [41]:
import pandas as pd
data = pd.read_csv('dtp_last.csv', sep=';')

In [42]:
data.head()

reg_code                регион    road_code  \
0        46    Московская область          NaN   
1        53  Оренбургская область  53К-2118000   
2        46    Московская область        А-107   
3         8      Хабаровский край        А-375   
4        35       Республика Крым      35Н-076   

                                              Дорога  \
0                       мкр. Авиационный-Лямцино-ММК   
1                                 Обход г. Оренбурга   
2  "Московское малое кольцо" Икша - Ногинск - Бро...   
3  "Восток" Хабаровск - Красный Яр - Ариадное - Ч...   
4             Симферополь - Севастополь - Бахчисарай   

                                           road_type  \
0  Региональная или межмуниципальная дорога общег...   
1  Региональная или межмуниципальная дорога общег...   
2              Федеральная дорога общего пользования   
3              Федеральная дорога общего пользования   
4  Региональная или межмуниципальная дорога общег...   

                                             address       crash_type_name  \
0  ЦФО, Московская область, Городской округ Домод...         Опрокидывание   
1                ПФО, Оренбургская область, Оренбург  Наезд на препятствие   
2         ЦФО, Московская область, Дмитровский район         Опрокидывание   
3           ДФО, Хабаровский край, Хабаровский район     Наезд на пешехода   
4         ЮФО, Республика Крым, Бахчисарайский район     Наезд на пешехода   

   crash_date crash_time                                       crash reason  \
0    20181231   23:55:00                    Другие нарушения ПДД водителями   
1    20181231   23:55:00                    Другие нарушения ПДД водителями   
2    20181231   23:40:00         Превышение установленной скорости движения   
3    20181231   23:20:00  Нахождение на проезжей части без цели её перех...   
4    20181231   23:00:00      Нарушение правил проезда пешеходного перехода   

   fatalities_amount  victims_amount  vehicles_amount participants_amount  
0                  1               0                1                  12  
1                  1               0                1                   2  
2                  1               0                1                   2  
3                  0               1                1                   2  
4                  1               0                1                   2

In [43]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39102 entries, 0 to 39101
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   reg_code             39102 non-null  int64 
 1   регион               39083 non-null  object
 2   road_code            20777 non-null  object
 3   Дорога               38910 non-null  object
 4   road_type            38910 non-null  object
 5   address              39102 non-null  object
 6   crash_type_name      39102 non-null  object
 7   crash_date           39102 non-null  int64 
 8   crash_time           39102 non-null  object
 9   crash reason         38467 non-null  object
 10  fatalities_amount    39102 non-null  int64 
 11  victims_amount       39102 non-null  int64 
 12  vehicles_amount      39102 non-null  int64 
 13  participants_amount  39102 non-null  object
dtypes: int64(5), object(9)
memory usage: 4.2+ MB


### Выводы:

1. Всего 39102 строки и 14 столбцов;
2. Заголовки на русском и английском языке, да еще и не в змеином регистре;
3. Дата в формате object;

Это пока все выводы, которые мы можем сделать просто взглянув на данные.

<div class="alert alert-block alert-success">
Ok<br>
<code>participants_amount</code> тоже object

### Предобработка заголовков

In [44]:
# помню, что желательно, чтобы заголовки были на английском языке, но проше прощения за плохой английский и оставляю все на русском, чтобы было легче ориентироваться
data.columns = ['код_региона', 'название_региона', 'код_дороги', 'название_дороги', 'тип_дороги', 'адрес', 'тип_ДТП', 'дата_ДТП', 'время_ДТП', 'причина_ДТП', 'число_погибших', 'число_пострадавших', 'число_транспортных_средств', 'число_участников']

In [45]:
# проверим
data.head()

код_региона      название_региона   код_дороги  \
0           46    Московская область          NaN   
1           53  Оренбургская область  53К-2118000   
2           46    Московская область        А-107   
3            8      Хабаровский край        А-375   
4           35       Республика Крым      35Н-076   

                                     название_дороги  \
0                       мкр. Авиационный-Лямцино-ММК   
1                                 Обход г. Оренбурга   
2  "Московское малое кольцо" Икша - Ногинск - Бро...   
3  "Восток" Хабаровск - Красный Яр - Ариадное - Ч...   
4             Симферополь - Севастополь - Бахчисарай   

                                          тип_дороги  \
0  Региональная или межмуниципальная дорога общег...   
1  Региональная или межмуниципальная дорога общег...   
2              Федеральная дорога общего пользования   
3              Федеральная дорога общего пользования   
4  Региональная или межмуниципальная дорога общег...   

                                               адрес               тип_ДТП  \
0  ЦФО, Московская область, Городской округ Домод...         Опрокидывание   
1                ПФО, Оренбургская область, Оренбург  Наезд на препятствие   
2         ЦФО, Московская область, Дмитровский район         Опрокидывание   
3           ДФО, Хабаровский край, Хабаровский район     Наезд на пешехода   
4         ЮФО, Республика Крым, Бахчисарайский район     Наезд на пешехода   

   дата_ДТП время_ДТП                                        причина_ДТП  \
0  20181231  23:55:00                    Другие нарушения ПДД водителями   
1  20181231  23:55:00                    Другие нарушения ПДД водителями   
2  20181231  23:40:00         Превышение установленной скорости движения   
3  20181231  23:20:00  Нахождение на проезжей части без цели её перех...   
4  20181231  23:00:00      Нарушение правил проезда пешеходного перехода   

   число_погибших  число_пострадавших  число_транспортных_средств  \
0               1                   0                           1   
1               1                   0                           1   
2               1                   0                           1   
3               0                   1                           1   
4               1                   0                           1   

  число_участников  
0               12  
1                2  
2                2  
3                2  
4                2

### Обработка пропущенных значений

In [46]:
# проверим количество пропусков в каждом столбце
data.isnull().sum()

код_региона                       0
название_региона                 19
код_дороги                    18325
название_дороги                 192
тип_дороги                      192
адрес                             0
тип_ДТП                           0
дата_ДТП                          0
время_ДТП                         0
причина_ДТП                     635
число_погибших                    0
число_пострадавших                0
число_транспортных_средств        0
число_участников                  0
dtype: int64

In [47]:
# начнем со столбца с названием региона
data[data['название_региона'].isna()]

код_региона название_региона код_дороги  \
36583           78              NaN        NaN   
36685           78              NaN        М-8   
36725           78              NaN        NaN   
36746           78              NaN        NaN   
36808           78              NaN        NaN   
36908           78              NaN        NaN   
37146           78              NaN        М-8   
37164           78              NaN        М-8   
37273           78              NaN        NaN   
37350           78              NaN        NaN   
37425           78              NaN        М-8   
37547           78              NaN        М-8   
37656           78              NaN        NaN   
37695           78              NaN        М-8   
37746           78              NaN        М-8   
37854           78              NaN        М-8   
37903           78              NaN        NaN   
37964           78              NaN        NaN   
38381           78              NaN        М-8   

                                         название_дороги  \
36583                         ПРЕЧИСТОЕ-КОЗА-СЕМЕНОВСКОЕ   
36685               Подъезд к городу Кострома от а/д М-8   
36725                      ЯРОСЛАВЛЬ-КРАСНЫЕ ТКАЧИ-ШОПША   
36746                         КРАСНЫЙ ПРОФИНТЕРН-ВЯТСКОЕ   
36808                                       ЛЮБИМ-ТРОИЦА   
36908                                  ЯРОСЛАВЛЬ-РЫБИНСК   
37146  "Холмогоры" Москва - Ярославль - Вологда – Арх...   
37164  "Холмогоры" Москва - Ярославль - Вологда – Арх...   
37273                              УГЛИЧ-НЕКОУЗ-БРЕЙТОВО   
37350             ОБХОД ЯРОСЛАВЛЯ С МОСТОМ ЧЕРЕЗ р,ВОЛГА   
37425  "Холмогоры" Москва - Ярославль - Вологда – Арх...   
37547  "Холмогоры" Москва - Ярославль - Вологда – Арх...   
37656     0км подъезд к Костроме - ул Калинина Ярославль   
37695               Подъезд к городу Кострома от а/д М-8   
37746                      Обход г. Переславль-Залесский   
37854  "Холмогоры" Москва - Ярославль - Вологда – Арх...   
37903                                  ЯРОСЛАВЛЬ-РЫБИНСК   
37964  СЕМИБРАТОВО-ВАХРУШЕВО-БАКЛАНОВО-Н.НИКОЛЬСКОЕ-С...   
38381  "Холмогоры" Москва - Ярославль - Вологда – Арх...   

                                              тип_дороги  \
36583  Региональная или межмуниципальная дорога общег...   
36685              Федеральная дорога общего пользования   
36725  Региональная или межмуниципальная дорога общег...   
36746  Региональная или межмуниципальная дорога общег...   
36808  Региональная или межмуниципальная дорога общег...   
36908  Региональная или межмуниципальная дорога общег...   
37146              Федеральная дорога общего пользования   
37164              Федеральная дорога общего пользования   
37273  Региональная или межмуниципальная дорога общег...   
37350  Региональная или межмуниципальная дорога общег...   
37425              Федеральная дорога общего пользования   
37547              Федеральная дорога общего пользования   
37656  Региональная или межмуниципальная дорога общег...   
37695              Федеральная дорога общего пользования   
37746              Федеральная дорога общего пользования   
37854              Федеральная дорога общего пользования   
37903  Региональная или межмуниципальная дорога общег...   
37964  Региональная или межмуниципальная дорога общег...   
38381              Федеральная дорога общего пользования   

                                                   адрес              тип_ДТП  \
36583       ЦФО, Ярославская область, Первомайский район        Опрокидывание   
36685        ЦФО, Ярославская область, Ярославский район         Столкновение   
36725        ЦФО, Ярославская область, Ярославский район         Столкновение   
36746       ЦФО, Ярославская область, Некрасовский район    Наезд на пешехода   
36808          ЦФО, Ярославская область, Любимский район  Наезд на стоящее ТС   
36908  ЦФО, Ярославская область, Ярославский район, Н...         Столкновение   
37146       ЦФО, Ярославская область,

In [48]:
# Все пропуски в названии региона под кодом 78. Спасибо гуглу, который подсказал, что 78 - это Ярославская область, которой мы и заполним все пропуски
data['название_региона'] = data['название_региона'].fillna('Ярославская область')

In [49]:
# проверка
data[data['название_региона'].isna()]

Empty DataFrame
Columns: [код_региона, название_региона, код_дороги, название_дороги, тип_дороги, адрес, тип_ДТП, дата_ДТП, время_ДТП, причина_ДТП, число_погибших, число_пострадавших, число_транспортных_средств, число_участников]
Index: []

In [50]:
# проверим столбец "код_дороги" на пропуски, возможно найдется закономероность
data[data['код_дороги'].isna()]

код_региона     название_региона код_дороги  \
0               46   Московская область        NaN   
6                1       Алтайский край        NaN   
8               56   Пензенская область        NaN   
12               7  Ставропольский край        NaN   
13               3   Краснодарский край        NaN   
...            ...                  ...        ...   
39085           63  Саратовская область        NaN   
39087           46   Московская область        NaN   
39095           46   Московская область        NaN   
39098           79    Республика Адыгея        NaN   
39101           63  Саратовская область        NaN   

                                         название_дороги  \
0                           мкр. Авиационный-Лямцино-ММК   
6                                Вторая Каменка - Гилево   
8      "Автодорога ""р.п. Беково - с. Варварино - р.п...   
12     Ачикулак - Махмуд-Мектеб Тукуй Мектеб 0000 - 4...   
13                  г.Армавир-с.Успенское-г.Невинномысск   
...                                                  ...   
39085  "автомобильная дорога ""Самара - Пугачев - Энг...   
39087         МКАД-Крутицы-Железнодорожный-Ликино-Дулево   
39095      Куровское-Шатура-Дмитровский Погост-Самойлиха   
39098                               Подъезд к а.Вочепший   
39101  "автомобильная дорога ""Самара - Пугачев - Энг...   

                                              тип_дороги  \
0      Региональная или межмуниципальная дорога общег...   
6      Региональная или межмуниципальная дорога общег...   
8      Региональная или межмуниципальная дорога общег...   
12     Региональная или межмуниципальная дорога общег...   
13     Региональная или межмуниципальная дорога общег...   
...                                                  ...   
39085  Региональная или межмуниципальная дорога общег...   
39087  Региональная или межмуниципальная дорога общег...   
39095  Региональная или межмуниципальная дорога общег...   
39098  Региональная или межмуниципальная дорога общег...   
39101  Региональная или межмуниципальная дорога общег...   

                                                   адрес            тип_ДТП  \
0      ЦФО, Московская область, Городской округ Домод...      Опрокидывание   
6      СФО, Алтайский край, Локтевский район, Гилевск...  Наезд на пешехода   
8      ПФО, Пензенская область, Бековский район, Верт...  Наезд на пешехода   
12     СКФО, Ставропольский край, Нефтекумский район,...      Опрокидывание   
13     ЮФО, Краснодарский край, Успенский район, Коно...       Столкновение   
...                                                  ...                ...   
39085        ПФО, Саратовская область, Энгельсский район       Столкновение   
39087           ЦФО, Московская область, Железнодорожный  Наезд на пешехода   
39095    ЦФО, Московская область, Орехово-Зуевский район       Столкновение   
39098           ЮФО, Республика Адыгея, Теучежский район  Наезд на животное   
39101        ПФО, Саратовская область, Энгельсский район       Столкновение   

       дата_ДТП время_ДТП                                        причина_ДТП  \
0      20181231  23:55:00                    Другие нарушения ПДД водителями   
6      20181231  21:40:00  Несоответствие скорости конкретным условиям дв...   
8      20181231  21:20:00  Переход через проезжую часть в неустановленном...   
12     20181231  21:00:00                    Другие нарушения ПДД водителями   
13     20181231  20:59:00                   Несоблюдение очередности проезда   
...         ...       ...                                                ...   
39085  20180401   0:05:00                Выезд на полосу встречного движения   
39087  20180401   4:15:00  Нахождение на проезжей части без цели её перехода   
39095  20180401   2:40:00                   Несоблюдение очередности проезда   
39098  20180401   1:30:00                    Другие нарушения ПДД водителями   
39101  20180401   0:05:00                Выезд на полосу встречного движения   

      

In [51]:
# к сожалению закономерность не найдено, поэтому просто заменим на слово неизвестно
data['код_дороги'] = data['код_дороги'].fillna('Неизвестно')

In [52]:
# проверка
data[data['код_дороги'].isna()]

Empty DataFrame
Columns: [код_региона, название_региона, код_дороги, название_дороги, тип_дороги, адрес, тип_ДТП, дата_ДТП, время_ДТП, причина_ДТП, число_погибших, число_пострадавших, число_транспортных_средств, число_участников]
Index: []

In [53]:
# проверим столбец "название_дороги" на пропуски, возможно найдется закономероность
data[data['название_дороги'].isna()]

код_региона       название_региона  код_дороги название_дороги  \
229             52         Омская область  Неизвестно             NaN   
347             87        Республика Коми  Неизвестно             NaN   
521             11  Архангельская область  Неизвестно             NaN   
549             22  Нижегородская область  Неизвестно             NaN   
672             11  Архангельская область  Неизвестно             NaN   
...            ...                    ...         ...             ...   
38158           66     Смоленская область  Неизвестно             NaN   
38306           22  Нижегородская область  Неизвестно             NaN   
38351           52         Омская область  Неизвестно             NaN   
38838           22  Нижегородская область  Неизвестно             NaN   
38923           22  Нижегородская область  Неизвестно             NaN   

      тип_дороги                                              адрес  \
229          NaN              СФО, Омская область, Знаменский район   
347          NaN          СЗФО, Республика Коми, Усть-Вымский район   
521          NaN   СЗФО, Архангельская область, Каргопольский район   
549          NaN  ПФО, Нижегородская область, Краснобаковский ра...   
672          NaN    СЗФО, Архангельская область, Холмогорский район   
...          ...                                                ...   
38158        NaN           ЦФО, Смоленская область, Велижский район   
38306        NaN      ПФО, Нижегородская область, Воротынский район   
38351        NaN                  СФО, Омская область, Омский район   
38838        NaN      ПФО, Нижегородская область, Володарский район   
38923        NaN       ПФО, Нижегородская область, Тонкинский район   

                    тип_ДТП  дата_ДТП время_ДТП  \
229            Столкновение  20181230  15:00:00   
347           Опрокидывание  20181229  23:30:00   
521       Наезд на пешехода  20181229   5:10:00   
549            Столкновение  20181228  22:15:00   
672    Наезд на препятствие  20181228  12:00:00   
...                     ...       ...       ...   
38158     Наезд на пешехода  20180412   4:20:00   
38306         Опрокидывание  20180410  11:30:00   
38351          Столкновение  20180409  18:10:00   
38838   Наезд на стоящее ТС  20180404   8:30:00   
38923          Столкновение  20180403   0:05:00   

                                             причина_ДТП  число_погибших  \
229    Нарушение правил расположения ТС на проезжей ч...               0   
347    Несоответствие скорости конкретным условиям дв...               1   
521        Неожиданный выход из-за сооружений (деревьев)               1   
549    Превышение установленной скорости движения, Вы...               3   
672                      Другие нарушения ПДД водителями               1   
...                                                  ...             ...   
38158                    Другие нарушения ПДД водителями               1   
38306  Нарушение правил расположения ТС на проезжей ч...               1   
38351                   Несоблюдение очередности проезда               2   
38838  Несоответствие скорости конкретным условиям дв...               1   
38923                Выезд на полосу встречного движения               0   

       число_пострадавших  число_транспортных_средств число_участников  
229                     1                           3                3  
347                     0                           1                2  
521                     0                           1                2  
549                     0                           2                4  
672                     0                           1               12  
...                   ...                         ...              ...  
38158                   0                           1                2  
38306                   0                           1               12  
38351                   0                           2                3  
38838                

In [54]:
# закономерность не найдена, поэтому просто заменим на слово неизвестно
data['код_дороги'] = data['код_дороги'].fillna('Неизвестно')

In [55]:
# проверим столбец "тип_дороги" на пропуски, возможно найдется закономероность
data[data['тип_дороги'].isna()]

код_региона       название_региона  код_дороги название_дороги  \
229             52         Омская область  Неизвестно             NaN   
347             87        Республика Коми  Неизвестно             NaN   
521             11  Архангельская область  Неизвестно             NaN   
549             22  Нижегородская область  Неизвестно             NaN   
672             11  Архангельская область  Неизвестно             NaN   
...            ...                    ...         ...             ...   
38158           66     Смоленская область  Неизвестно             NaN   
38306           22  Нижегородская область  Неизвестно             NaN   
38351           52         Омская область  Неизвестно             NaN   
38838           22  Нижегородская область  Неизвестно             NaN   
38923           22  Нижегородская область  Неизвестно             NaN   

      тип_дороги                                              адрес  \
229          NaN              СФО, Омская область, Знаменский район   
347          NaN          СЗФО, Республика Коми, Усть-Вымский район   
521          NaN   СЗФО, Архангельская область, Каргопольский район   
549          NaN  ПФО, Нижегородская область, Краснобаковский ра...   
672          NaN    СЗФО, Архангельская область, Холмогорский район   
...          ...                                                ...   
38158        NaN           ЦФО, Смоленская область, Велижский район   
38306        NaN      ПФО, Нижегородская область, Воротынский район   
38351        NaN                  СФО, Омская область, Омский район   
38838        NaN      ПФО, Нижегородская область, Володарский район   
38923        NaN       ПФО, Нижегородская область, Тонкинский район   

                    тип_ДТП  дата_ДТП время_ДТП  \
229            Столкновение  20181230  15:00:00   
347           Опрокидывание  20181229  23:30:00   
521       Наезд на пешехода  20181229   5:10:00   
549            Столкновение  20181228  22:15:00   
672    Наезд на препятствие  20181228  12:00:00   
...                     ...       ...       ...   
38158     Наезд на пешехода  20180412   4:20:00   
38306         Опрокидывание  20180410  11:30:00   
38351          Столкновение  20180409  18:10:00   
38838   Наезд на стоящее ТС  20180404   8:30:00   
38923          Столкновение  20180403   0:05:00   

                                             причина_ДТП  число_погибших  \
229    Нарушение правил расположения ТС на проезжей ч...               0   
347    Несоответствие скорости конкретным условиям дв...               1   
521        Неожиданный выход из-за сооружений (деревьев)               1   
549    Превышение установленной скорости движения, Вы...               3   
672                      Другие нарушения ПДД водителями               1   
...                                                  ...             ...   
38158                    Другие нарушения ПДД водителями               1   
38306  Нарушение правил расположения ТС на проезжей ч...               1   
38351                   Несоблюдение очередности проезда               2   
38838  Несоответствие скорости конкретным условиям дв...               1   
38923                Выезд на полосу встречного движения               0   

       число_пострадавших  число_транспортных_средств число_участников  
229                     1                           3                3  
347                     0                           1                2  
521                     0                           1                2  
549                     0                           2                4  
672                     0                           1               12  
...                   ...                         ...              ...  
38158                   0                           1                2  
38306                   0                           1               12  
38351                   0                           2                3  
38838                

In [56]:
# закономерность не найдена, поэтому просто заменим на слово неизвестно
data['тип_дороги'] = data['тип_дороги'].fillna('Неизвестно')

In [57]:
# проверим столбец "тип_дороги" на пропуски, возможно найдется закономероность
data[data['причина_ДТП'].isna()]

код_региона                          название_региона  код_дороги  \
56             718  Ханты-Мансийский автономный округ — Югра  Неизвестно   
308             84                          Республика Алтай       Р-256   
375             80                   Республика Башкортостан  Неизвестно   
392             29                         Калужская область  Неизвестно   
533             52                            Омская область       Р-402   
...            ...                                       ...         ...   
38765           45                                    Москва       А-109   
38775           99              Еврейская автономная область       Р-297   
38946           11                     Архангельская область     11К-612   
38950           35                           Республика Крым     35К-013   
39065           58                         Псковская область     58К-019   

                                         название_дороги  \
56                        а/д г.Сургут - г.Нижневартовск   
308    "Чуйский тракт" Новосибирск - Барнаул - Горно-...   
375    Арсланово-Кушнаренково(Дмитриевка,Бишкази,Ст.Т...   
392                                        Калуга-Медынь   
533                    Тюмень - Ялуторовск - Ишим - Омск   
...                                                  ...   
38765                                    Ильинское шоссе   
38775  "Амур" Чита - Невер - Свободный - Архара - Бир...   
38946  Карпогоры - Сосновка - Нюхча - граница с Респу...   
38950                           Черноморское - Евпатория   
39065                             Пушкинские Горы- Локня   

                                              тип_дороги  \
56     Региональная или межмуниципальная дорога общег...   
308                Федеральная дорога общего пользования   
375    Региональная или межмуниципальная дорога общег...   
392    Региональная или межмуниципальная дорога общег...   
533                Федеральная дорога общего пользования   
...                                                  ...   
38765              Федеральная дорога общего пользования   
38775              Федеральная дорога общего пользования   
38946  Региональная или межмуниципальная дорога общег...   
38950  Региональная или межмуниципальная дорога общег...   
39065  Региональная или межмуниципальная дорога общег...   

                                                   адрес  \
56     УФО, Ханты-Мансийский автономный округ, Нижнев...   
308             СФО, Республика Алтай, Шебалинский район   
375    ПФО, Республика Башкортостан, Кушнаренковский ...   
392            ЦФО, Калужская область, Дзержинский район   
533                СФО, Омская область, Саргатский район   
...                                                  ...   
38765  ЦФО, Москва, Участки обслуживания на территори...   
38775  ДФО, Еврейская автономная область, Смидовичски...   
38946       СЗФО, Архангельская область, Пинежский район   
38950           ЮФО, Республика Крым, Черноморский район   
39065  СЗФО, Псковская область, Новоржевский район, В...   

                    тип_ДТП  дата_ДТП время_ДТП причина_ДТП  число_погибших  \
56            Опрокидывание  20181231  15:00:00         NaN               1   
308           Опрокидывание  20181230   9:00:00         NaN               1   
375       Наезд на пешехода  20181229  19:30:00         NaN               0   
392            Столкновение  20181229  18:30:00         NaN               1   
533           Опрокидывание  20181229   0:30:00         NaN               1   
...                     ...       ...       ...         ...             ...   
38765     Падение пассажира  20180405  11:10:00         NaN               1   
38775         Опрокидывание  20180405   7:20:00         NaN               2   
38946          Столкновение  20180402  16:30:00         NaN               1   
38950         Опрокидывание  20180402  15:05:00         NaN               1   
39065  Наезд на препятствие  20180401   7:10:00         NaN   

In [58]:
# закономерность не найдена, поэтому просто заменим на слово неизвестно
data['причина_ДТП'] = data['причина_ДТП'].fillna('Неизвестно')

<div class="alert alert-block alert-success">
Согласен, пропуски заполнены верно!

### Поиск проблем в данных

In [59]:
# спустя долгих поисков проблемы, я наткнулась на такое интересное наблюдение: все числа числятся как строки, да  еще и строка больше 25
data['число_участников'].unique()

array(['12', '2', '5', '3', '4', '6', '8', '10', '7', '14', '9', '24',
       '16', '11', '13', '15', 'больше 25', '22', '20', '18', '19', '17',
       '23'], dtype=object)

In [60]:
# исправим больше 25 на 25 и переведем в числовой формат
data['число_участников'] = data['число_участников'].replace('больше 25', '25')
data['число_участников'] = data['число_участников'].astype(int)

In [61]:
data['число_участников'].unique()

array([12,  2,  5,  3,  4,  6,  8, 10,  7, 14,  9, 24, 16, 11, 13, 15, 25,
       22, 20, 18, 19, 17, 23])

In [62]:
# еще одна проблема: неявные дубликаты из-за слов обл. и области, еще и регионов из-за этого 119
print(data['название_региона'].unique())
data['название_региона'].nunique()

['Московская область' 'Оренбургская область' 'Хабаровский край'
 'Республика Крым' 'Республика Ингушетия' 'Алтайский край'
 'Республика Башкортостан' 'Пензенская область' 'Республика Адыгея'
 'Ставропольский край' 'Краснодарский край' 'Республика Тыва'
 'Пермский край' 'Чувашская Республика — Чувашия' 'Камчатский край'
 'Иркутская область' 'Новосибирская область' 'Калужская область'
 'Кировская область' 'Еврейская автономная область'
 'Архангельская область' 'Тульская область' 'Владимирская область'
 'Нижегородская область' 'Астраханская область' 'Ивановская область'
 'Ярославская область' 'Удмуртская Республика' 'Вологодская область'
 'Красноярский край' 'Ростовская область' 'Республика Калмыкия'
 'Ханты-Мансийский автономный округ — Югра' 'Воронежская область'
 'Тюменская область' 'Мурманская область' 'Ленинградская область'
 'Карачаево-Черкесская Республика' 'Тверская область'
 'Республика Марий Эл' 'Республика Мордовия' 'Приморский край'
 'Республика Бурятия' 'Республика Коми' 'Смо

119

In [63]:
# исправим это недоразумение
data['название_региона'] = data['название_региона'].str.replace('обл\.', 'область')

C:\Users\User\AppData\Local\Temp\ipykernel_62120\3840244191.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  data['название_региона'] = data['название_региона'].str.replace('обл\.', 'область')


In [64]:
# а еще Лен область лучше переименовать в Ленинградскую область
data['название_региона'] = data['название_региона'].replace(to_replace=['Лен.обл.'], value=['Ленинградская область'], regex=True)

In [66]:
# проверочка
print(data['название_региона'].unique())
data['название_региона'].nunique()

['Московская область' 'Оренбургская область' 'Хабаровский край'
 'Республика Крым' 'Республика Ингушетия' 'Алтайский край'
 'Республика Башкортостан' 'Пензенская область' 'Республика Адыгея'
 'Ставропольский край' 'Краснодарский край' 'Республика Тыва'
 'Пермский край' 'Чувашская Республика — Чувашия' 'Камчатский край'
 'Иркутская область' 'Новосибирская область' 'Калужская область'
 'Кировская область' 'Еврейская автономная область'
 'Архангельская область' 'Тульская область' 'Владимирская область'
 'Нижегородская область' 'Астраханская область' 'Ивановская область'
 'Ярославская область' 'Удмуртская Республика' 'Вологодская область'
 'Красноярский край' 'Ростовская область' 'Республика Калмыкия'
 'Ханты-Мансийский автономный округ — Югра' 'Воронежская область'
 'Тюменская область' 'Мурманская область' 'Ленинградская область'
 'Карачаево-Черкесская Республика' 'Тверская область'
 'Республика Марий Эл' 'Республика Мордовия' 'Приморский край'
 'Республика Бурятия' 'Республика Коми' 'Смо

84

<div class="alert alert-block alert-success">
Отлично! Все проблемы найдены

### Преобразование типов данных

In [67]:
# число участников лучше оставить в численном виде
data['число_участников'] = data['число_участников'].astype(int)

In [68]:
# дату перевели в соотвествующий формат
data['дата_ДТП'] = pd.to_datetime(data['дата_ДТП'], format='%Y%m%d')

In [69]:
# теперь все в порядке
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39102 entries, 0 to 39101
Data columns (total 14 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   код_региона                 39102 non-null  int64         
 1   название_региона            39102 non-null  object        
 2   код_дороги                  39102 non-null  object        
 3   название_дороги             38910 non-null  object        
 4   тип_дороги                  39102 non-null  object        
 5   адрес                       39102 non-null  object        
 6   тип_ДТП                     39102 non-null  object        
 7   дата_ДТП                    39102 non-null  datetime64[ns]
 8   время_ДТП                   39102 non-null  object        
 9   причина_ДТП                 39102 non-null  object        
 10  число_погибших              39102 non-null  int64         
 11  число_пострадавших          39102 non-null  int64     

<div class="alert alert-block alert-success">
Верно!

### Поиск дубликатов

In [70]:
# итак, найдем явные дубликаты
data.duplicated().sum()

17

In [71]:
data[data.duplicated()]

код_региона         название_региона  код_дороги  \
37227           22    Нижегородская область    22К-0016   
39086           61        Рязанская область         М-5   
39087           46       Московская область  Неизвестно   
39088           61        Рязанская область        Р-22   
39089           27  Калининградская область     27А-007   
39090           32      Кемеровская область       Р-255   
39091           61        Рязанская область       Р-132   
39092           86       Республика Карелия       А-121   
39093           56       Пензенская область         М-5   
39094           18    Волгоградская область        Р-22   
39095           46       Московская область  Неизвестно   
39096           94    Удмуртская Республика         М-7   
39097           80  Республика Башкортостан     80Н-069   
39098           79        Республика Адыгея  Неизвестно   
39099            3       Краснодарский край       Р-217   
39100           76       Забайкальский край       А-350   
39101           63      Саратовская область  Неизвестно   

                                         название_дороги  \
37227                                Семенов - Ковернино   
39086  "Урал" Москва - Рязань - Пенза - Самара - Уфа ...   
39087         МКАД-Крутицы-Железнодорожный-Ликино-Дулево   
39088  «Каспий» автомобильная дорога М-4 "Дон" - Тамб...   
39089                     Северный обход г. Калининграда   
39090                Подъезд к городу Томск от а/д Р-255   
39091                  Калуга - Тула - Михайлов - Рязань   
39092  "Сортавала" Санкт-Петербург - Сортавала - авто...   
39093  "Урал" Москва - Рязань - Пенза - Самара - Уфа ...   
39094  «Каспий» автомобильная дорога М-4 "Дон" - Тамб...   
39095      Куровское-Шатура-Дмитровский Погост-Самойлиха   
39096                  Подъезд к городу Пермь от а/д М-7   
39097                   Аскарово - Альмухаметово - Сибай   
39098                               Подъезд к а.Вочепший   
39099               Подъезд к городу Майкоп от а/д Р-217   
39100  Чита - Забайкальск - граница с Китайской Народ...   
39101  "автомобильная дорога ""Самара - Пугачев - Энг...   

                                              тип_дороги  \
37227  Региональная или межмуниципальная дорога общег...   
39086              Федеральная дорога общего пользования   
39087  Региональная или межмуниципальная дорога общег...   
39088              Федеральная дорога общего пользования   
39089  Региональная или межмуниципальная дорога общег...   
39090              Федеральная дорога общего пользования   
39091              Федеральная дорога общего пользования   
39092              Федеральная дорога общего пользования   
39093              Федеральная дорога общего пользования   
39094              Федеральная дорога общего пользования   
39095  Региональная или межмуниципальная дорога общег...   
39096              Федеральная дорога общего пользования   
39097  Региональная или межмуниципальная дорога общег...   
39098  Региональная или межмуниципальная дорога общег...   
39099              Федеральная дорога общего пользования   
39100              Федеральная дорога общего пользования   
39101  Региональная или межмуниципальная дорога общег...   

                                                   адрес  \
37227  ПФО, Нижегородская область, Городской округ Се...   
39086               ЦФО, Рязанская область, Шацкий район   
39087           ЦФО, Московская область, Железнодорожный   
39088         ЦФО, Рязанская область, Михайловский район   
39089    СЗФО, Калининградская область, Гурьевский район   
39090          СФО, Кемеровская область, Юргинский район   
39091    ЦФО, Рязанская область, Рязань, Железнодорожный   
39092  СЗФО, Республика Карелия, Лахденпохский район,...   
39093  ПФО, Пензенская область, Кузнецкий район, Евла...   
39094      ЮФО, Волгоградская область, Иловлинский район   
39095    ЦФО, Московская область, Орехово-Зуевский район   
39096   ПФО, Удмуртская Республика, Малопургинский район   
39

In [72]:
# на всякий случай еще один способ
data[data.duplicated() == True]

код_региона         название_региона  код_дороги  \
37227           22    Нижегородская область    22К-0016   
39086           61        Рязанская область         М-5   
39087           46       Московская область  Неизвестно   
39088           61        Рязанская область        Р-22   
39089           27  Калининградская область     27А-007   
39090           32      Кемеровская область       Р-255   
39091           61        Рязанская область       Р-132   
39092           86       Республика Карелия       А-121   
39093           56       Пензенская область         М-5   
39094           18    Волгоградская область        Р-22   
39095           46       Московская область  Неизвестно   
39096           94    Удмуртская Республика         М-7   
39097           80  Республика Башкортостан     80Н-069   
39098           79        Республика Адыгея  Неизвестно   
39099            3       Краснодарский край       Р-217   
39100           76       Забайкальский край       А-350   
39101           63      Саратовская область  Неизвестно   

                                         название_дороги  \
37227                                Семенов - Ковернино   
39086  "Урал" Москва - Рязань - Пенза - Самара - Уфа ...   
39087         МКАД-Крутицы-Железнодорожный-Ликино-Дулево   
39088  «Каспий» автомобильная дорога М-4 "Дон" - Тамб...   
39089                     Северный обход г. Калининграда   
39090                Подъезд к городу Томск от а/д Р-255   
39091                  Калуга - Тула - Михайлов - Рязань   
39092  "Сортавала" Санкт-Петербург - Сортавала - авто...   
39093  "Урал" Москва - Рязань - Пенза - Самара - Уфа ...   
39094  «Каспий» автомобильная дорога М-4 "Дон" - Тамб...   
39095      Куровское-Шатура-Дмитровский Погост-Самойлиха   
39096                  Подъезд к городу Пермь от а/д М-7   
39097                   Аскарово - Альмухаметово - Сибай   
39098                               Подъезд к а.Вочепший   
39099               Подъезд к городу Майкоп от а/д Р-217   
39100  Чита - Забайкальск - граница с Китайской Народ...   
39101  "автомобильная дорога ""Самара - Пугачев - Энг...   

                                              тип_дороги  \
37227  Региональная или межмуниципальная дорога общег...   
39086              Федеральная дорога общего пользования   
39087  Региональная или межмуниципальная дорога общег...   
39088              Федеральная дорога общего пользования   
39089  Региональная или межмуниципальная дорога общег...   
39090              Федеральная дорога общего пользования   
39091              Федеральная дорога общего пользования   
39092              Федеральная дорога общего пользования   
39093              Федеральная дорога общего пользования   
39094              Федеральная дорога общего пользования   
39095  Региональная или межмуниципальная дорога общег...   
39096              Федеральная дорога общего пользования   
39097  Региональная или межмуниципальная дорога общег...   
39098  Региональная или межмуниципальная дорога общег...   
39099              Федеральная дорога общего пользования   
39100              Федеральная дорога общего пользования   
39101  Региональная или межмуниципальная дорога общег...   

                                                   адрес  \
37227  ПФО, Нижегородская область, Городской округ Се...   
39086               ЦФО, Рязанская область, Шацкий район   
39087           ЦФО, Московская область, Железнодорожный   
39088         ЦФО, Рязанская область, Михайловский район   
39089    СЗФО, Калининградская область, Гурьевский район   
39090          СФО, Кемеровская область, Юргинский район   
39091    ЦФО, Рязанская область, Рязань, Железнодорожный   
39092  СЗФО, Республика Карелия, Лахденпохский район,...   
39093  ПФО, Пензенская область, Кузнецкий район, Евла...   
39094      ЮФО, Волгоградская область, Иловлинский район   
39095    ЦФО, Московская область, Орехово-Зуевский район   
39096   ПФО, Удмуртская Республика, Малопургинский район   
39

In [84]:
# просто удалим эти 17 дубликатов, при этом поменяем индекс после наших преобразований
data.drop_duplicates(inplace=True)
data = data.drop_duplicates().reset_index(drop=True)

<div class="alert alert-block alert-success">
Дубликаты уничтожены

### Добавление столбцов.

In [77]:
# объеденим наши два столбца даты и времении в один
data['дата_и_время'] = data['дата_ДТП'].astype(str) + 'T' + data['время_ДТП']
data['дата_и_время'] = pd.to_datetime(data['дата_и_время'], format='%Y-%m-%d %H:%M:%S')

In [78]:
# сначала создала столбец с днями неделями на английском
data['день_недели'] = data['дата_ДТП'].dt.day_name()

In [79]:
dict = {'Monday' : 'Понедельник',
        'Tuesday' : 'Вторник',
        'Wednesday' : 'СРеда',
        'Thursday' : 'Четверг',
        'Friday' : 'Пятница',
        'Saturday' : 'Суббота',
        'Sunday' : 'Воскресение'}

In [80]:
# с помощью словаря перевела все на русский язык + проверочка
data['день_недели_на_русском'] = data['день_недели'].map(dict)
data.head()

код_региона      название_региона   код_дороги  \
0           46    Московская область   Неизвестно   
1           53  Оренбургская область  53К-2118000   
2           46    Московская область        А-107   
3            8      Хабаровский край        А-375   
4           35       Республика Крым      35Н-076   

                                     название_дороги  \
0                       мкр. Авиационный-Лямцино-ММК   
1                                 Обход г. Оренбурга   
2  "Московское малое кольцо" Икша - Ногинск - Бро...   
3  "Восток" Хабаровск - Красный Яр - Ариадное - Ч...   
4             Симферополь - Севастополь - Бахчисарай   

                                          тип_дороги  \
0  Региональная или межмуниципальная дорога общег...   
1  Региональная или межмуниципальная дорога общег...   
2              Федеральная дорога общего пользования   
3              Федеральная дорога общего пользования   
4  Региональная или межмуниципальная дорога общег...   

                                               адрес               тип_ДТП  \
0  ЦФО, Московская область, Городской округ Домод...         Опрокидывание   
1                ПФО, Оренбургская область, Оренбург  Наезд на препятствие   
2         ЦФО, Московская область, Дмитровский район         Опрокидывание   
3           ДФО, Хабаровский край, Хабаровский район     Наезд на пешехода   
4         ЮФО, Республика Крым, Бахчисарайский район     Наезд на пешехода   

    дата_ДТП время_ДТП                                        причина_ДТП  \
0 2018-12-31  23:55:00                    Другие нарушения ПДД водителями   
1 2018-12-31  23:55:00                    Другие нарушения ПДД водителями   
2 2018-12-31  23:40:00         Превышение установленной скорости движения   
3 2018-12-31  23:20:00  Нахождение на проезжей части без цели её перех...   
4 2018-12-31  23:00:00      Нарушение правил проезда пешеходного перехода   

   число_погибших  число_пострадавших  число_транспортных_средств  \
0               1                   0                           1   
1               1                   0                           1   
2               1                   0                           1   
3               0                   1                           1   
4               1                   0                           1   

   число_участников день_недели        дата_и_время день_недели_на_русском  
0                12      Monday 2018-12-31 23:55:00            Понедельник  
1                 2      Monday 2018-12-31 23:55:00            Понедельник  
2                 2      Monday 2018-12-31 23:40:00            Понедельник  
3                 2      Monday 2018-12-31 23:20:00            Понедельник  
4                 2      Monday 2018-12-31 23:00:00            Понедельник

<div class="alert alert-block alert-success">
Мой любимый способ, хотя можно импортировать локаль )

In [83]:
# все шикарно, теперь удалим ненужные столбцы!
data.drop(data.columns[[7,8, 14]], axis=1)

код_региона         название_региона   код_дороги  \
0               46       Московская область   Неизвестно   
1               53     Оренбургская область  53К-2118000   
2               46       Московская область        А-107   
3                8         Хабаровский край        А-375   
4               35          Республика Крым      35Н-076   
...            ...                      ...          ...   
39097           80  Республика Башкортостан      80Н-069   
39098           79        Республика Адыгея   Неизвестно   
39099            3       Краснодарский край        Р-217   
39100           76       Забайкальский край        А-350   
39101           63      Саратовская область   Неизвестно   

                                         название_дороги  \
0                           мкр. Авиационный-Лямцино-ММК   
1                                     Обход г. Оренбурга   
2      "Московское малое кольцо" Икша - Ногинск - Бро...   
3      "Восток" Хабаровск - Красный Яр - Ариадное - Ч...   
4                 Симферополь - Севастополь - Бахчисарай   
...                                                  ...   
39097                   Аскарово - Альмухаметово - Сибай   
39098                               Подъезд к а.Вочепший   
39099               Подъезд к городу Майкоп от а/д Р-217   
39100  Чита - Забайкальск - граница с Китайской Народ...   
39101  "автомобильная дорога ""Самара - Пугачев - Энг...   

                                              тип_дороги  \
0      Региональная или межмуниципальная дорога общег...   
1      Региональная или межмуниципальная дорога общег...   
2                  Федеральная дорога общего пользования   
3                  Федеральная дорога общего пользования   
4      Региональная или межмуниципальная дорога общег...   
...                                                  ...   
39097  Региональная или межмуниципальная дорога общег...   
39098  Региональная или межмуниципальная дорога общег...   
39099              Федеральная дорога общего пользования   
39100              Федеральная дорога общего пользования   
39101  Региональная или межмуниципальная дорога общег...   

                                                   адрес  \
0      ЦФО, Московская область, Городской округ Домод...   
1                    ПФО, Оренбургская область, Оренбург   
2             ЦФО, Московская область, Дмитровский район   
3               ДФО, Хабаровский край, Хабаровский район   
4             ЮФО, Республика Крым, Бахчисарайский район   
...                                                  ...   
39097  ПФО, Республика Башкортостан, Абзелиловский ра...   
39098           ЮФО, Республика Адыгея, Теучежский район   
39099       ЮФО, Краснодарский край, Новокубанский район   
39100  СФО, Забайкальский край, Читинский район, Ново...   
39101        ПФО, Саратовская область, Энгельсский район   

                    тип_ДТП  \
0             Опрокидывание   
1      Наезд на препятствие   
2             Опрокидывание   
3         Наезд на пешехода   
4         Наезд на пешехода   
...                     ...   
39097     Наезд на животное   
39098     Наезд на животное   
39099          Столкновение   
39100         Опрокидывание   
39101          Столкновение   

                                             причина_ДТП  число_погибших  \
0                        Другие нарушения ПДД водителями               1   
1                        Другие нарушения ПДД водителями               1   
2             Превышение установленной скорости движения               1   
3      Нахождение на проезжей части без цели её перех...               0   
4          Нарушение правил проезда пешеходного перехода               1   
...                                                  ...             ...   
39097  Нарушение правил расположения ТС на проезжей ч...               2   
39098                    Другие нарушения ПДД водителями               0   
39099  Выезд на полосу встречного движения в местах, ...          

### Экспорт результатов

In [88]:
data.to_excel('prikol.xlsx', index=False)

### Выводы:

1. Были проблемы с Ярославской областью. Возможно там какие-то неполадки, либо она на какое-то время пропала и не высвечилась в базе данных, но мы все исправили!
2. Дубликатов было 17, все были успешно ликвидированы.
3. Разобрались с облс. и областями и теперь их 84. Тем кто составлял эти данные нужно договориться и ввести единый стандарт по заполнению.
4. Для более удобного использование объеденили дату и время, еще и столбец добавили с днем недели. Теперь можно легко построить график и установить зависимость, например, между днем неделей и любым другим признаком из таблицы.